packages/sw-background-sync-queue/src/lib/requestManager.js

```diff
@@ -0,0 +1,136 @@
+import {initializationDefaults} from './constants';
+import queue from './queue.js';
+import responseManager from './responseManager';
+let globalConfig = initializationDefaults;
+let globalCallbacks = {};
+
+/**
+ * Class to handle all the request related
+ * transformations, replaying, event handling
+ * broadcasting back to controlled pages etc.
+ * @class RequestManager
+ */
+class RequestManager {
+	/**
+	 * Initializes the request manager
+	 * stores the callbacks object, maintins global config and
+	 * attaches event handler
+	 * @param {Object} config config that gives values like
+	 * maxAge of requests in iDB
+	 * @param {Object} callbacks object that contains function
+	 * ref for various lifecycle callbacks
+	 *
+	 * @memberOf RequestManager
+	 */
+	async initialize(config, callbacks) {
+		this.attachSyncHandler();
+		globalConfig = Object.assign({}, initializationDefaults, config);
+		globalCallbacks = callbacks;
+		await queue.initialize( globalConfig );
+		await queue.cleanupQueue();
+	}
+
+	/**
+	 * attaches sync handler to replay requests when
+	 * sync event is fired
+	 *
+	 * @memberOf RequestManager
+	 */
+	attachSyncHandler() {
+		self.addEventListener('sync', (event) => {
+			event.waitUntil(this.replayRequests());
+		});
+	}
+
+	/**
+	 * function to start playing requests
+	 * in sequence
+	 * @return {Promise} promise that will resolve whenn all requests complete
+	 *
+	 * @memberOf RequestManager
+	 */
+	replayRequests() {
+		if (queue.getTotalTasks() < 1) {
+			return;
+		}
+		return this.doFetch(0);
+	}
+
+	/**
+	 * takes starting index and play all then
+	 * requests from that index onwards till the end
+	 *
+	 * @param {int} index starting point of playing
+	 * @return {void}
+	 *
+	 * @memberOf RequestManager
+	 */
+	async doFetch(index) {
+		let reqData = await queue.getRequestFromQueueAtIndex(index);
+		let hash = queue.getHash(index);
+		// exit point
+		if(!reqData) {
+			return;
+		}
+
+		// proceed if response do not already exist
+		if(reqData.response) {
+			return this.doFetch(index + 1);
+		}
+
+		let request = new Request(reqData.request.url, {
+			method: reqData.request.method,
+			headers: JSON.parse(reqData.request.headers),
+		});
+
+		if(request.method === 'POST') {
+			request.body = reqData.request.body;
```

Comment:
I think you'll either want to do some type checking against the data (is it binary?) or have some sanitization before trying to assign otherwise it will fail.